server/models/meshmodel/registration/dir.go

```diff
@@ -0,0 +1,89 @@
+package registration
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+
+	"github.com/layer5io/meshkit/models/meshmodel/core/v1alpha2"
+	"github.com/layer5io/meshkit/models/meshmodel/core/v1beta1"
+	"github.com/layer5io/meshkit/models/meshmodel/entity"
+)
+
+type Dir struct {
+	dirpath string
+}
+
+/*
+	The directory should contain one and only one `model`.
+	A directory containing multiple `model` will be invalid.
+*/
+func NewDir(path string) Dir {
+	return Dir{dirpath: path}
+}
+
+
+/* PkgUnit parses all the files inside the directory and finds out if they are any valid meshery definitions. Valid meshery definitions are added to the packagingUnit struct.
+*/
+func (d Dir) PkgUnit() (_ packagingUnit, err error){
+	pkg := packagingUnit{}
+	// check if the given is a directory
+	_, err = os.ReadDir(d.dirpath)
+	if(err != nil){
+		return pkg, ErrDirPkgUnitParseFail(d.dirpath, fmt.Errorf("Could not read the directory: %e", err))
+	}
+	err = filepath.Walk(d.dirpath, func (path string, f os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if(f.IsDir()){
+				return nil
+		}
+		byt, _ := os.ReadFile(path)
+		var e entity.Entity
+		if(filepath.Ext(path) == ".yaml"){
```

Comment:
Why? Is ".yml" not valid?

Why is the file extension a concern at all? Unless there's an identifiable justification, this approach is fragile and over-optimized.

Reply:
Yes, I thought of using the syntax to identify the file type. But I thought depending on the error type from the `json` and `yaml` packages would be fragile as they often have less detail. Look at this - https://stackoverflow.com/questions/68156758/robust-way-check-if-the-input-is-json-or-yaml , the first answer gives me the impression that it might be fragile, especially with YAML files.

I am not feeling good about this approach either, but I thought this would not be an issue, because at this stage, we already know that we are dealing with a directory, and all the standard parsers check for file extension as the first check before checking anything else.

Reply:
I just am not entirely confident with the alternate approach.

Reply:
Call me when you have stable Internet and I'll explain why figuring out whether it's json or yaml upfront, simply doesn't matter.

Reply:
Just attempt to parse the file. Apply a JSON validator, if valid continue. If JSON fail, apply a YAML validator. If good, continue. If fail, exit.

The filename extension is indicative of which validator to run, and as an optimization, we can run the validator appropriate to the filename extension first, however, ultimately, we’re going to have to try to parse the file whether it has a filename extension or not.

Reply:
@Yashs911 @Jougan-0 fyi

Reply:
https://github.com/meshery/meshery/pull/11385/files#diff-2c08829638b3e8aca48699c241f868611fede1f71e1523eb5ac924f307f31553R13 - a generalised `unmarshal` function has been introduced, which unmarshals the entity irrespective of its file type.

Reply:
The registration logic has been updated to make use of this function without checking for filetypes.